#1396 multiple curve options

georgeDaskalakis opened the pull request:
closes #1396

georgeDaskalakis commented on src/OSPSuite.Presentation/Presenters/Charts/CurveMultiItemEditorPresenter.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace OSPSuite.Presentation.Presenters.Charts
+{
+   public enum YesNoValues
```
@msevestre no particularly proud of this solution here....still I preferred Yes/No instead of true/false and the enum keeps things consistent with the other comboboxes.

msevestre replied:
Why do we need this? Why is our yes no not enough?

msevestre commented on src/OSPSuite.Assets/UIConstants.cs:
```diff
@@ -1253,6 +1255,16 @@ public static class ColorGrouping
+         public static class MultiCurveOptions
+         {
+            public static string CurrentValue = "<Current value>";
+            public static string Color = "Color";
```
We have those constants somewhere. Please reuse. They clearly mean the same

georgeDaskalakis replied:
done

msevestre commented on src/OSPSuite.Presentation/Presenters/Charts/CurveMultiItemEditorPresenter.cs:
```diff
@@ -0,0 +1,50 @@
+   {
+      public Color? Color { get; set; }
+      public string Style { get; set; }
+      public string Symbol { get; set; }
```
Why is everything designed as string?we have specific enums etc..this is very suspicious

georgeDaskalakis replied:
@msevestre let us briefly talk about this...I used to have them as nullable enum, but was not very fond of that behaviour either.

msevestre commented on src/OSPSuite.Presentation/Presenters/Charts/CurveSettingsPresenter.cs:
```diff
@@ -203,5 +208,32 @@ public void UpdateColorForCurve(Curve curve, Color color)
+            foreach (var curveDTO in selectedCurveDTOs)
+            {
+               if (selectedValues.Color != null) curveDTO.Color = (Color)selectedValues.Color;
+               if (Enum.TryParse(selectedValues.Style, out LineStyles selectedStyle)) 
```
Lol..yes this was very suspicious. 
Just set the value using real types (maybe nullable) instead or all those cast. 
Unless there is something that I don't understand? In that case I need some explanations

msevestre commented on src/OSPSuite.Presentation/Presenters/Charts/CurveSettingsPresenter.cs:
```diff
@@ -203,5 +208,32 @@ public void UpdateColorForCurve(Curve curve, Color color)
+      {
+         using (var multiEditorPresenter = _applicationController.Start<ICurveMultiItemEditorPresenter>())
+         {
+            multiEditorPresenter.Show();
```
Just encapsulate this behaviour. You are calling 3 presenter methods in a row. 
Call only one and put the logic in the presenter

msevestre reviewed (changes requested):
Lots of things that I do not understand:
* Usage of a new enum type for yes/no when we have our types already
* Usage of string as type for everything! From boolean to enums.

Let's chat

georgeDaskalakis commented:
@msevestre I changed this and implemented with the screenBinder as we discussed. Please take a look of the changes and if you have no objections feel free to merge :)

msevestre commented on src/OSPSuite.Presentation/Presenters/Charts/CurveMultiItemEditorPresenter.cs:
```diff
@@ -0,0 +1,57 @@
+   public class CurveMultiItemEditorPresenter : AbstractDisposablePresenter<ICurveMultiItemEditorView, ICurveMultiItemEditorPresenter>,
+      ICurveMultiItemEditorPresenter
+   {
+      private SelectedCurveValues _selectedCurveValues;
```
Run resharper on this file. It's clearly not formatter properly. Let's run resharper on all files before we submit a PR

georgeDaskalakis replied:
good point, I do go through the changes, but just by looking at the diff it is easy to lose something.

msevestre commented on src/OSPSuite.Presentation/Formatters/MultiCurveOptionsFormatter.cs:
```diff
@@ -0,0 +1,39 @@
+         if (valueToFormat == null)
+            return Captions.Chart.MultiCurveOptions.CurrentValue;
+
+         return  valueToFormat.Value ? "Yes" : "No";
```
@georgeDaskalakis no need to cast. use value

msevestre commented on src/OSPSuite.Presentation/Presenters/Charts/CurveMultiItemEditorPresenter.cs:
```diff
@@ -0,0 +1,59 @@
+
+      public IEnumerable<bool?> AllBooleanOptions { get; } = Constants.MultiCurveOptions.AllBooleanOptions;
+
+      public IEnumerable<LineStyles?> AllLineStyles { get; } = new List<LineStyles?>() {null}.Union(EnumHelper.AllValuesFor<LineStyles>().Cast<LineStyles?>());
```
no need to evaluate the propertie all the time. Define a single property

georgeDaskalakis replied:
good point

msevestre commented on src/OSPSuite.Presentation/Presenters/Charts/CurveMultiItemEditorPresenter.cs:
```diff
@@ -0,0 +1,59 @@
+
+      public void SetColorChangedFlag()
+      {
+         _selectedCurveValues.IsColorSet = true;
```
I don't get this IsColorSet property. Why not Color?

georgeDaskalakis replied:
so the Color? did not work with binding to the ColorEdit Control. Also not with the formatter, so the best workaround I could think of was to use the flag.

msevestre commented on src/OSPSuite.Presentation/Presenters/Charts/CurveSettingsPresenter.cs:
```diff
@@ -203,5 +208,26 @@ public void UpdateColorForCurve(Curve curve, Color color)
+               if (selectedValues.IsColorSet)
+                  curveDTO.Color = selectedValues.Color;
+
+               curveDTO.LineStyle = selectedValues.Style.GetValueOrDefault(curveDTO.LineStyle);
```
use the properties of nullable

msevestre commented on src/OSPSuite.UI/Views/Charts/CurveMultiItemEditorView.cs:
```diff
@@ -0,0 +1,84 @@
+      private readonly IFormatter<bool?> _boolFormatter = new BooleanYesNoFormatter();
+      private readonly IFormatter<LineStyles?> _lineStylesFormatter = new LineStylesFormatter();
+      private readonly IFormatter<Symbols?> _symbolsFormatter = new SymbolsFormatter();
+      private readonly ScreenBinder<SelectedCurveValues> _screenBinder = new ScreenBinder<SelectedCurveValues> {BindingMode = BindingMode.TwoWay};
```
define as readonly in cosntructor or direct init

msevestre commented on src/OSPSuite.UI/Views/Charts/CurveMultiItemEditorView.Designer.cs:
```diff
@@ -0,0 +1,266 @@
+         {
+            components.Dispose();
+         }
+         _screenBinder.Dispose();
```
Very important to dispoise of binder otherwise you have a leak

msevestre commented on src/OSPSuite.Presentation/Presenters/Charts/CurveMultiItemEditorPresenter.cs:
```diff
@@ -0,0 +1,59 @@
+      public Symbols? Symbol { get; set; }
+      public bool? Visible { get; set; }
+      public bool? VisibleInLegend { get; set; }
+      public bool IsColorSet { get; set; } = false;
```
@georgeDaskalakis I do not understand this flag

georgeDaskalakis replied:
@msevestre thanks for all the other points and corrections! Again this flag has to do with not being able to bind the Color? property to the ColorEdit Control.

msevestre replied:
I see. Probably we could add support for Color? in DevExpressBinder. Merging for now

msevestre reviewed (commented):
I have updated the code to match the coding standard and ran resharper. Let's try to do it before submitting the PR
Also: Why do we need the IsColorSet flag? Color? not enough i guess but I dont understand why
